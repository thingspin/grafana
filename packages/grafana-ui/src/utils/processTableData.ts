import { TableData, Column } from '../types/index';

import Papa, { ParseError, ParseMeta } from 'papaparse';

// Subset of all parse options
export interface TableParseOptions {
  headerIsFirstLine?: boolean; // Not a papa-parse option
  delimiter?: string; // default: ","
  newline?: string; // default: "\r\n"
  quoteChar?: string; // default: '"'
  encoding?: string; // default: ""
  comments?: boolean | string; // default: false
}

export interface TableParseDetails {
  meta?: ParseMeta;
  errors?: ParseError[];
}

/**
 * This makes sure the header and all rows have equal length.
 *
 * @param table (immutable)
 * @returns a new table that has equal length rows, or the same
 * table if no changes were needed
 */
export function matchRowSizes(table: TableData): TableData {
  const { rows } = table;
  let { columns } = table;

  let sameSize = true;
  let size = columns.length;
  rows.forEach(row => {
    if (size !== row.length) {
      sameSize = false;
      size = Math.max(size, row.length);
    }
  });
  if (sameSize) {
    return table;
  }

  // Pad Columns
  if (size !== columns.length) {
    const diff = size - columns.length;
    columns = [...columns];
    for (let i = 0; i < diff; i++) {
      columns.push({
        text: 'Column ' + (columns.length + 1),
      });
    }
  }

  // Pad Rows
  const fixedRows: any[] = [];
  rows.forEach(row => {
    const diff = size - row.length;
    if (diff > 0) {
      row = [...row];
      for (let i = 0; i < diff; i++) {
        row.push(null);
      }
    }
    fixedRows.push(row);
  });

  return {
    columns,
    rows: fixedRows,
    type: table.type,
    columnMap: table.columnMap,
  };
}

function makeColumns(values: any[]): Column[] {
  return values.map((value, index) => {
    if (!value) {
      value = 'Column ' + (index + 1);
    }
    return {
      text: value.toString().trim(),
    };
  });
}

/**
 * Convert CSV text into a valid TableData object
 *
 * @param text
 * @param options
 * @param details, if exists the result will be filled with debugging details
 */
export function parseCSV(text: string, options?: TableParseOptions, details?: TableParseDetails): TableData {
  const results = Papa.parse(text, { ...options, dynamicTyping: true, skipEmptyLines: true });
  const { data, meta, errors } = results;

  // Fill the parse details for debugging
  if (details) {
    details.errors = errors;
    details.meta = meta;
  }

  if (!data || data.length < 1) {
    // Show a more reasonable warning on empty input text
    if (details && !text) {
      errors.length = 0;
      errors.push({
        code: 'empty',
        message: 'Empty input text',
        type: 'warning',
        row: 0,
      });
      details.errors = errors;
    }
    return {
      columns: [],
      rows: [],
      type: 'table',
      columnMap: {},
    };
  }

  // Assume the first line is the header unless the config says its not
  const headerIsNotFirstLine = options && options.headerIsFirstLine === false;
  const header = headerIsNotFirstLine ? [] : results.data.shift();

  return matchRowSizes({
    columns: makeColumns(header),
    rows: results.data,
    type: 'table',
    columnMap: {},
  });
}
