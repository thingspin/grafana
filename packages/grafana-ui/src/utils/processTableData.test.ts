import { parseCSV } from './processTableData';

describe('processTableData', () => {
  describe('basic processing', () => {
    it('should read header and two rows', () => {
      const text = 'a,b,c\n1,2,3\n4,5,6';
      expect(parseCSV(text)).toMatchSnapshot();
    });

    it('should generate a header and fix widths', () => {
      const text = '1\n2,3,4\n5,6';
      const table = parseCSV(text, {
        headerIsFirstLine: false,
      });
      expect(table.rows.length).toBe(3);

      expect(table).toMatchSnapshot();
    });
  });
});
