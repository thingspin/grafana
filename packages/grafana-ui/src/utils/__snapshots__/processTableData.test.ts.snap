// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processTableData basic processing should generate a header and fix widths 1`] = `
Object {
  "columnMap": Object {},
  "columns": Array [
    Object {
      "text": "Column 1",
    },
    Object {
      "text": "Column 2",
    },
    Object {
      "text": "Column 3",
    },
  ],
  "rows": Array [
    Array [
      1,
      null,
      null,
    ],
    Array [
      2,
      3,
      4,
    ],
    Array [
      5,
      6,
      null,
    ],
  ],
  "type": "table",
}
`;

exports[`processTableData basic processing should read header and two rows 1`] = `
Object {
  "columnMap": Object {},
  "columns": Array [
    Object {
      "text": "a",
    },
    Object {
      "text": "b",
    },
    Object {
      "text": "c",
    },
  ],
  "rows": Array [
    Array [
      1,
      2,
      3,
    ],
    Array [
      4,
      5,
      6,
    ],
  ],
  "type": "table",
}
`;
